app/routines/generate_reset_code.rb

```diff
@@ -0,0 +1,18 @@
+class GenerateResetCode
```

Comment:
You are right that you get a good refactoring by making `ResetCode` its own class, but I think this might be a reasonable example of when a routine is overkill, since only internal state of ResetCode is involved.  

Also, `ResetCode` reads like a routine, where "reset" sounds like a verb.  Since these are only used for password resets, can we rename it "password reset code"?

I'd be in favor of 

```
class Identity
  ...
  def reset_password_reset_code!(blah)
    self.build_password_reset_code && self.save if password_reset_code.nil?
    self.password_reset_code.reset!(blah)
  end
end

class PasswordResetCode
  ...
  def reset!(blah)
    # set values
    save
  end
end
```

and then a call to `identity.reset_password_reset_code!(blah)`.

That seem reasonable?